app/integration_action_test.go

```diff
@@ -68,250 +68,282 @@ func TestPostActionInvalidURL(t *testing.T) {
 }
 
 func TestPostAction(t *testing.T) {
-	th := Setup(t).InitBasic()
-	defer th.TearDown()
-
-	th.App.UpdateConfig(func(cfg *model.Config) {
-		*cfg.ServiceSettings.AllowedUntrustedInternalConnections = "localhost,127.0.0.1"
-	})
-
-	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		request := model.PostActionIntegrationRequestFromJson(r.Body)
-		assert.NotNil(t, request)
-
-		assert.Equal(t, request.UserId, th.BasicUser.Id)
-		assert.Equal(t, request.UserName, th.BasicUser.Username)
-		assert.Equal(t, request.ChannelId, th.BasicChannel.Id)
-		assert.Equal(t, request.ChannelName, th.BasicChannel.Name)
-		assert.Equal(t, request.TeamId, th.BasicTeam.Id)
-		assert.Equal(t, request.TeamName, th.BasicTeam.Name)
-		assert.True(t, len(request.TriggerId) > 0)
-		if request.Type == model.POST_ACTION_TYPE_SELECT {
-			assert.Equal(t, request.DataSource, "some_source")
-			assert.Equal(t, request.Context["selected_option"], "selected")
-		} else {
-			assert.Equal(t, request.DataSource, "")
-		}
-		assert.Equal(t, "foo", request.Context["s"])
-		assert.EqualValues(t, 3, request.Context["n"])
-		fmt.Fprintf(w, `{"post": {"message": "updated"}, "ephemeral_text": "foo"}`)
-	}))
-	defer ts.Close()
+	testCases := []struct {
+		Description string
+		Channel     func(th *TestHelper) *model.Channel
+	}{
+		{"public channel", func(th *TestHelper) *model.Channel {
+			return th.BasicChannel
+		}},
+		{"direct channel", func(th *TestHelper) *model.Channel {
+			user1 := th.CreateUser()
+
+			return th.CreateDmChannel(user1)
+		}},
+		{"group channel", func(th *TestHelper) *model.Channel {
+			user1 := th.CreateUser()
+			user2 := th.CreateUser()
+
+			return th.CreateGroupChannel(user1, user2)
+		}},
+	}
 
-	interactivePost := model.Post{
-		Message:       "Interactive post",
-		ChannelId:     th.BasicChannel.Id,
-		PendingPostId: model.NewId() + ":" + fmt.Sprint(model.GetMillis()),
-		UserId:        th.BasicUser.Id,
-		Props: model.StringInterface{
-			"attachments": []*model.SlackAttachment{
-				{
-					Text: "hello",
-					Actions: []*model.PostAction{
+	for _, testCase := range testCases {
+		t.Run(testCase.Description, func(t *testing.T) {
+			th := Setup(t).InitBasic()
+			defer th.TearDown()
+
+			channel := testCase.Channel(th)
+
+			th.App.UpdateConfig(func(cfg *model.Config) {
+				*cfg.ServiceSettings.AllowedUntrustedInternalConnections = "localhost,127.0.0.1"
+			})
+
+			ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				request := model.PostActionIntegrationRequestFromJson(r.Body)
+				assert.NotNil(t, request)
+
+				assert.Equal(t, request.UserId, th.BasicUser.Id)
+				assert.Equal(t, request.UserName, th.BasicUser.Username)
+				assert.Equal(t, request.ChannelId, channel.Id)
+				assert.Equal(t, request.ChannelName, channel.Name)
+				if channel.Type == model.CHANNEL_DIRECT || channel.Type == model.CHANNEL_GROUP {
```

Comment:
The diff here is painful, but the only meaningful change is here, plus parameterizing the channel throughout the tests.